examples/routing_service/udp_socket_adapter/src/SocketStreamReader.cxx

```diff
@@ -0,0 +1,184 @@
+        /**
+         * Essential to protect against concurrent data access to
+         * buffer_ from the take() methods running on a different
+         * Routing Service thread.
+         */
+        std::unique_lock<std::mutex> lock(buffer_mutex_);
+        socket->receive_data(
+                received_buffer_,
+                &received_bytes_,
+                BUFFER_MAX_SIZE);
+        lock.unlock();
+
+        // Most likely received nothing or there was an error
+        // Not doing any error handling here
+        if (received_bytes_ <= 0) {
+            // Sleep for a small period of time to avoid busy waiting
+            std::this_thread::sleep_for(std::chrono::milliseconds(10));
+            continue;
+        }
+
+        /**
+         * Here we notify Routing Service, that there is data available
+         * on the StreamReader, triggering a call to take().
+         */
+        reader_listener_->on_data_available(this);
+    }
+
+    socket_connection_->dispose_discovery_stream(stream_info_);
+}
+
+SocketStreamReader::SocketStreamReader(
+        SocketConnection *connection,
+        const StreamInfo &info,
+        const PropertySet &properties,
+        StreamReaderListener *listener)
+        : stop_thread_(false),
+          stream_info_(info.stream_name(), info.type_info().type_name())
+{
+    socket_connection_ = connection;
+    reader_listener_ = listener;
+    adapter_type_ =
+            static_cast<DynamicType *>(info.type_info().type_representation());
+
+    // Parse the properties provided in the xml configuration file
+    for (const auto &property : properties) {
+        if (property.first == RECEIVE_ADDRESS_STRING) {
+            receive_address_ = property.second;
+        } else if (property.first == RECEIVE_PORT_STRING) {
+            receive_port_ = std::stoi(property.second);
+        } else if (property.first == SHAPE_COLOR_STRING) {
+            shape_color_ = property.second;
+        }
+    }
+
+    // If any of the mandatory properties is not specified, throw exception
+    if (receive_address_.size() == 0 || receive_port_ == 0
+        || shape_color_.size() == 0) {
+        throw dds::core::IllegalOperationError(
+                "You must set receive_address, receive_port and"
+                " shape_color in the RsSocketAdapter.xml file");
+    }
+
+    // Create the UDP socket to receive data
+    socket = std::unique_ptr<UdpSocket>(
+            new UdpSocket(receive_address_.c_str(), receive_port_));
+
+    // Start the receive thread for UDP data
+    socketreader_thread_ =
+            std::thread(&SocketStreamReader::socket_reading_thread, this);
+}
+
+/**
+ * This is the Routing Service take(). It's called when the
+ * socket_receive_thread calls on_data_available()
+ */
+void SocketStreamReader::take(
+        std::vector<dds::core::xtypes::DynamicData *> &samples,
+        std::vector<dds::sub::SampleInfo *> &infos)
+{
+    if (stream_info_.stream_name() == "Square") {
```

Comment:
Comment here or where it can be relevant, that the code is hard-coded to work with Square.

Reply:
Good point. I went ahead and added this information in the XML file, and changed this code to also support Circle and Triangle in case people want to use it